### PR TITLE
feat: Added autocomplete for setters in  Input, JSONForm and Table

### LINKED ISSUE #22708: Phase 2 [Task] Support autocomplete for widget setter methods

### PR DESCRIPTION
**Pull request by Druthi**
## Description
Added autocomplete for setters in  Input, JSONForm and Table

#### PR fixes following issue(s)
Fixes #22708 

#### Type of change
- New feature (non-breaking change which adds functionality)

- [ ] Manual
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Test-plan-implementation#speedbreaker-features-to-consider-for-every-change) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans/_edit#areas-of-interest)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts [hunk 1]
```diff
@@ -216,3 +226,16 @@ export function generateJSFunctionTypeDef(
...
+
+export function addSettersToDefinitions(
+  definitions: Record<string, any>,
+  entityConfig?: WidgetEntityConfig,
+) {
+  if (entityConfig && entityConfig.__setters) {
+    const setters = Object.keys(entityConfig.__setters);
+
+    setters.forEach((setter: string) => {
+      definitions[setter] = "fn()";
```
Comment (Rishabh-Rathod) [inline app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts:238]: We need to think about how we would support the value type here.

Comment (Rishabh-Rathod) [reply app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts:238]: something like `fn(text: string )`

Comment (Druthi) [reply app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts:238]: Ahh this is what i was asking about. How would I get the `type` here?

File: app/client/src/entities/DataTree/dataTreeWidget.ts [hunk 2]
```diff
@@ -71,6 +71,7 @@ function getSetterConfig(
       for (const setterMethodName of Object.keys(setterConfig.__setters)) {
         modifiedSetterConfig.__setters[setterMethodName] = {
           path: `${widget.widgetName}.${setterConfig.__setters[setterMethodName].path}`,
+          type: setterConfig.__setters[setterMethodName].type,
```
Comment (Rishabh-Rathod) [inline app/client/src/entities/DataTree/dataTreeWidget.ts:74]: Do we need to send type in unEvallTree? 
Validation would run according to the validationPaths config correct?

File: app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts [hunk 3]
```diff
@@ -234,8 +234,10 @@ export function addSettersToDefinitions(
   if (entityConfig && entityConfig.__setters) {
     const setters = Object.keys(entityConfig.__setters);
 
-    setters.forEach((setter: string) => {
-      definitions[setter] = "fn()";
+    setters.forEach((setterName: string) => {
+      const setterType = entityConfig.__setters?.[setterName].type;
+
+      definitions[setterName] = `fn(value:${setterType})`;
```
Comment (Rishabh-Rathod) [inline app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts:240]: Shall we also add a config for the parameter name to have 
`fn(${propertyName}:${setterType})`;

**Review by Rishabh-Rathod — APPROVED** [review]

File: app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts (hunk 3 above)
Comment (Druthi) [reply app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts:240]: i suggest we don't. Setters core functionality is to set a value to a property. We would over complicate it by adding propertyName. I don't see a use case for this in the future.

File: app/client/src/entities/DataTree/dataTreeWidget.ts (hunk 2 above)
Comment (Druthi) [reply app/client/src/entities/DataTree/dataTreeWidget.ts:74]: I use the configTree to get the `path` and `type` in autocomplete and so I added it here.

**Pull request merged**